### DISCUSSION TITLE
Change RocksDB to 7.х?

### DISCUSSION THREAD

**Discussion by aleksraiden** [apache/kvrocks#1013] [General]
So, after we are switched to C++17, we are ready to use an 7.х family of RocksDB. Maybe we can try to use 7.0.4 - as a latest version without any changes in public API sections, and then step-by-step migrate to newest versions? 

I will submit PR, of course

**Comment by git-hulk** [comment]
Thanks for @aleksraiden proposal, I think it'd be better to wait for more services using 7.x in production, or it also makes sense if we need requests which appears only in 7.x.

**Comment by PragmaTwice** [comment]
Hi @aleksraiden ,
I notice there are lots of public API and behavior changes in rocksdb 7 (https://github.com/facebook/rocksdb/blob/main/HISTORY.md), will these affect kvrocks?
I am also curious about the performance of kvrocks between rocksdb 6 and 7, are there some benchmark about it?
For update to rocksdb 7 (if we decide to do that), I prefer update to the latest version directly.

**Comment by git-hulk** [comment]
https://groups.google.com/g/rocksdb/c/umRPboha0as/m/5VP6zvwiCgAJ?utm_medium=email&utm_source=footer

RocksDB community mentioned that 7.5.3 could significantly improve compaction time, so it may become one of the reasons to upgrade to 7.x, record it here first.

**Comment by xiaobiaozhao** [comment]
I think if we plan to upgrade to 7.x,  as well just upgrade to the latest version
Here's a demo pr
https://github.com/xiaobiaozhao/incubator-kvrocks/pull/4
I plan to do some test for performance improvements about rocksdb 7 vs rocksdb 6 at the weekend.

**Reply by git-hulk** [reply]
It will help a lot if you can provide more information about the 7.x performance improvement.

**Reply by git-hulk** [reply]
@aleksraiden Do you test the performance of the latest RocksDB? From RocksDB's release logs and group discussion, they said 7.x will significantly improve bulk writing and compaction.

**Comment by xiaobiaozhao** [comment]
## kvrocks with rocksdb_v6 vs rocksdv_v7 report
### Build
build current kvrocks and use rocksdb-v7 kvrocks  pr is  https://github.com/xiaobiaozhao/incubator-kvrocks/pull/4

### Test steps
Test deploy

![image](https://user-images.githubusercontent.com/52393536/198837849-6ed79fb7-e00f-468f-b4b7-ebde8d24a9db.png)

stress script
```
docker run -d --rm redislabs/memtier_benchmark:latest  -t 2 -c 20 -s 172.27.255.230 -p 6666 --distinct-client-seed  --key-minimum=100000000 --key-maximum=10000000000 --random-data  --data-size=10000 -n 10000000000 --command "set __key__ __data__"
```

kvrocks hardware settings
```
Cores:4C
RAM:32GB
Disk: nvme SSD 890GB
```

### monitor

> 172.27.255.230 -> kvrocks-v7
>
> 172.27.255.229 -> kvrocks-v6

kvrocks-v7 vs  kvrocks-v6 full compact monitor
```
127.0.0.1:6666> COMPACT
OK
```
![image](https://user-images.githubusercontent.com/52393536/198836052-fde65b46-47b2-4165-9a8d-deb3a706ce62.png)

172.27.255.229 disk monitor

![image](https://user-images.githubusercontent.com/52393536/198836721-f5a34ef1-aa81-4cbe-8261-0a98e6e6cd6a.png)

172.27.255.230 disk monitor

![image](https://user-images.githubusercontent.com/52393536/198836763-7bc44f8c-e82f-4487-bc6f-17b8ee07aff3.png)

Other monitoring measures were consistent

### When kvrocks compacting
|         | kvrocks(v6)                                     | kvrocks(v7)       |
| ------- | ----------------------------------------------- | ----------------- |
| cpu     | to lower                                        | normal            |
| mem     | to lower then to higher                         | normal            |
| disk    | reads doubled and limited 500MB（kvrocks.conf） | a little big high |
| ops     | to almost zero                                  | normal            |
| latency | to higher                                       | normal            |

### Conclusion
1. I think the v7 version has more consistent performance than v6 in compact
2. I think kvrocks worth upgrading to rocksdb v7

**Reply by git-hulk** [reply]
Thanks for @xiaobiaozhao great benchmark report, it's truly worth upgrading in bulk writing scenarios.

It looks can solve your issues. cc @juancrescente

**Reply by LucifierMT** [reply]
How to generate the picture like this、please tell me，thanks

**Reply by xiaobiaozhao** [reply]
which pic?

**Comment by mapleFU** [comment]
Here is another patch in RocksDB 7.6:
https://github.com/facebook/rocksdb/commit/1e9bf25f61fa5b08fe2b3fe5c1fcf33e4c1c515c
Which makes tasks faster

**Reply by git-hulk** [reply]
Thanks for @mapleFU input.
